File: java/ans/0108.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
TreeNode to(int[]nums,int beg,int end){
    if(beg>end)return null;
    int mid=(beg+end)/2;
    TreeNode ans=new TreeNode(nums[mid]);
    ans.left=to(nums,beg,mid-1);
    ans.right=to(nums,mid+1,end);
    return ans;
}
public TreeNode sortedArrayToBST(int[] nums) {
    return to(nums,0,nums.length-1);
}
}

File: rust/src/bin/0108.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;


use std::rc::Rc;
use std::cell::RefCell;

fn go(a: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
    if a.len() == 0 {
        return None;
    }
    let mid = a.len() / 2;
    let left = go(&a[..mid]);
    let right = go(&a[mid + 1..]);
    let mut node = TreeNode::new(a[mid]);
    node.left = left;
    node.right = right;
    return Some(Rc::new(RefCell::new(node)));
}

impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        go(&nums)
    }
}

fn main() {
    println!("good");
    Solution::sorted_array_to_bst(vec!(1, 2, 3, 4, 5));
}